src/core/ComboBox/ComboBox.tsx

```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        ? itemRenderer(option, {
+            isFocused: false,
+            isSelected: false,
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
 
-    const _filteredOptions =
-      filterFunction?.(options, inputValue) ??
-      options.filter((option) =>
-        option.label.toLowerCase().includes(inputValue?.trim().toLowerCase()),
+      return customItem ? (
+        React.cloneElement(customItem, {
+          onClick: (e: unknown) => {
+            dispatch(['select', __originalIndex as number]);
+            customItem.props.onClick?.(e);
+          },
+          'data-iui-index': __originalIndex,
+        })
+      ) : (
+        <ComboBoxMenuItem
+          key={(__originalIndex as number) ?? index}
+          index={option.__originalIndex}
+          onClick={() => dispatch(['select', __originalIndex as number])}
+          {...rest}
+        >
+          {option.label}
+        </ComboBoxMenuItem>
       );
-    setFilteredOptions(_filteredOptions);
+    });
+  }, [itemRenderer, options]);
 
-    setFocusedIndex((previouslyFocusedIndex) => {
-      if (_filteredOptions.includes(options[previouslyFocusedIndex])) {
-        return previouslyFocusedIndex;
-      } else if (
-        _filteredOptions.find(({ value }) => value === selectedValue)
+  const getMenuItem = React.useCallback(
+    (option: ComboBoxOption<T>) => {
+      const { __originalIndex, ...rest } = option;
+
+      if (
+        selectedIndex !== __originalIndex &&
+        focusedIndex !== __originalIndex
       ) {
-        return options.findIndex(({ value }) => value === selectedValue);
-      } else {
-        return -1; // reset focus if previously focused or selected value is not in filtered list
+        return memoizedItems[option.__originalIndex];
       }
-    });
-  }, [inputValue, options, selectedValue, isOpen, filterFunction]);
-
-  const onKeyDown = React.useCallback(
-    (event: React.KeyboardEvent) => {
-      const focusableOptions = getFocusableElements(menuRef.current);
-      const focusedIndexInFilteredList = focusableOptions.findIndex(
-        ({ id }) =>
-          id === inputRef.current?.getAttribute('aria-activedescendant'),
+
+      const customItem = itemRenderer
+        ? itemRenderer(option, {
+            isFocused: focusedIndex === __originalIndex,
+            isSelected: selectedIndex === __originalIndex,
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
+
+      return customItem ? (
+        React.cloneElement(customItem, {
+          onClick: (e: unknown) => {
+            dispatch(['select', __originalIndex as number]);
+            customItem.props.onClick?.(e);
+          },
+          // ComboBox.MenuItem handles data-iui-index and iui-focused through context
+          // but we still need to pass them here for backwards compatibility with MenuItem
+          className: cx(customItem.props.className, 'iui-focused'),
```

Comment:
you need to add `iui-focused` to all elements? doesnt make sense

Reply:
i have an early return above
<img width="509" alt="image" src="https://user-images.githubusercontent.com/9084735/163568680-0a7648e6-6a85-4a34-8989-6576f04fb25c.png">